src/chains/ethereum/ethereum/src/blockchain.ts

```diff
@@ -647,6 +668,73 @@ export default class Blockchain extends Emittery<BlockchainTypedEvents> {
     return vm;
   };
 
+  applySimulationOverrides = async (vm: VM, overrides: CallOverrides) => {
+    const stateManager = vm.stateManager;
+    for (const address in overrides) {
+      if (!hasOwn(overrides, address)) continue;
+      const { balance, nonce, code, state, stateDiff } = overrides[address];
+
+      const vmAddr = { buf: Address.from(address).toBuffer() } as any;
+      // group together overrides that update the account
+      if (nonce || balance || code) {
+        const account = await stateManager.getAccount(vmAddr);
+
+        if (nonce) {
+          account.nonce = {
+            toArrayLike: () => Quantity.from(nonce).toBuffer()
+          } as any;
+        }
+        if (balance) {
+          account.balance = {
+            toArrayLike: () => Quantity.from(balance).toBuffer()
+          } as any;
+        }
+        if (code) {
+          const codeBuffer = Data.from(code).toBuffer();
+          // The ethereumjs-vm StateManager does not allow to set empty code,
+          // therefore we will manually set the code hash when "clearing" the contract code
+          const codeHash =
+            codeBuffer.length > 0 ? keccak(codeBuffer) : KECCAK256_NULL;
+          account.codeHash = codeHash;
+          await this.trie.db.put(codeHash, codeBuffer);
+        }
+        await stateManager.putAccount(vmAddr, account);
+      }
+      // group together overrides that update storage
+      if (state || stateDiff) {
+        let newState: { [slot: string]: string };
+        if (state) {
+          // it's possible that the user fed an override with a valid address
+          // and slot, but not a value we can actually set in the storage. if
+          // so, we don't want to set the storage, and we also don't want to
+          // clear it out
+          let hasValidValue = false;
+          for (const slot in state) {
+            const value = state[slot];
+            if (!hasOwn(state, slot) || value == null || value === "") continue;
+            hasValidValue = true;
+          }
+          if (!hasValidValue) return;
+          // override.state clears all storage and sets just the specified slots
+          await stateManager.clearContractStorage(vmAddr);
+          newState = state;
+        } else {
+          newState = stateDiff;
+        }
+        for (const slot in newState) {
+          const value = newState[slot];
+          // don't set storage for invalid values
+          if (!hasOwn(newState, slot) || value == null || value === "")
```

Comment:
I think we only need to do this check for `stateDiff`, as the `state` branch already does this. Should be just remove the check from the `state` section? Alternatively, since we're already looping over all data in `state` in order to create the `newState` should we `putContractStorage` then, instead of here?

Reply:
I've reworked this section to be more efficient.